In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import functions as fun

from transform import Processor

Importing data

In [2]:
data = pd.read_csv('/Users/zwouq/Downloads/archive 2/GlobalLandTemperaturesByMajorCity.csv') 

#Descriptive statistics

In [3]:
data

dt  AverageTemperature  AverageTemperatureUncertainty  \
0       1849-01-01              26.704                          1.435   
1       1849-02-01              27.434                          1.362   
2       1849-03-01              28.101                          1.612   
3       1849-04-01              26.140                          1.387   
4       1849-05-01              25.427                          1.200   
...            ...                 ...                            ...   
239172  2013-05-01              18.979                          0.807   
239173  2013-06-01              23.522                          0.647   
239174  2013-07-01              25.251                          1.042   
239175  2013-08-01              24.528                          0.840   
239176  2013-09-01                 NaN                            NaN   

           City        Country Latitude Longitude  
0       Abidjan  Côte D'Ivoire    5.63N     3.23W  
1       Abidjan  Côte D'Ivoire    5.63N     3.23W  
2       Abidjan  Côte D'Ivoire    5.63N     3.23W  
3       Abidjan  Côte D'Ivoire    5.63N     3.23W  
4       Abidjan  Côte D'Ivoire    5.63N     3.23W  
...         ...            ...      ...       ...  
239172     Xian          China   34.56N   108.97E  
239173     Xian          China   34.56N   108.97E  
239174     Xian          China   34.56N   108.97E  
239175     Xian          China   34.56N   108.97E  
239176     Xian          China   34.56N   108.97E  

[239177 rows x 7 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 239177 entries, 0 to 239176
Data columns (total 7 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   dt                             239177 non-null  object 
 1   AverageTemperature             228175 non-null  float64
 2   AverageTemperatureUncertainty  228175 non-null  float64
 3   City                           239177 non-null  object 
 4   Country                        239177 non-null  object 
 5   Latitude                       239177 non-null  object 
 6   Longitude                      239177 non-null  object 
dtypes: float64(2), object(5)
memory usage: 12.8+ MB


Mean temperature by city and country

In [5]:
grouping_result = data.groupby(["Country", "City"])["AverageTemperature"].mean()
grouping_result.head(30)

Country      City          
Afghanistan  Kabul             14.342919
Angola       Luanda            23.693046
Australia    Melbourne         13.374304
             Sydney            17.002299
Bangladesh   Dhaka             25.490568
Brazil       Belo Horizonte    21.071396
             Brasília          21.727595
             Fortaleza         27.008640
             Rio De Janeiro    23.788916
             Salvador          24.656971
             São Paulo         19.699368
Burma        Rangoon           26.735193
Canada       Montreal           4.445014
             Toronto            5.773911
Chile        Santiago           5.692277
China        Changchun          4.923798
             Chengdu           10.638042
             Chongqing         16.844962
             Dalian            10.229077
             Guangzhou         21.608684
             Harbin             3.625744
             Jinan             13.089814
             Nanjing           15.602052
             Peking          

In [6]:
cities_wo_dupl = data.drop_duplicates(subset=["Latitude", "Longitude"])
country_grouping = cities_wo_dupl["Country"].value_counts()
print(country_grouping)

China                                 15
India                                 12
Brazil                                 6
Nigeria                                3
Turkey                                 3
United States                          3
Canada                                 2
South Africa                           2
Indonesia                              2
Saudi Arabia                           2
Russia                                 2
Pakistan                               2
Colombia                               2
Japan                                  2
Egypt                                  2
Australia                              2
Somalia                                1
Mexico                                 1
Côte D'Ivoire                          1
Italy                                  1
Kenya                                  1
France                                 1
Burma                                  1
Philippines                            1
Chile           

In [7]:
data.groupby("City")["dt"].agg([min, max])

min         max
City                               
Abidjan      1849-01-01  2013-09-01
Addis Abeba  1850-01-01  2013-09-01
Ahmadabad    1796-01-01  2013-09-01
Aleppo       1791-05-01  2013-09-01
Alexandria   1791-05-01  2013-09-01
...                 ...         ...
Tokyo        1845-01-01  2013-09-01
Toronto      1743-11-01  2013-09-01
Umm Durman   1859-01-01  2013-09-01
Wuhan        1841-01-01  2013-09-01
Xian         1820-08-01  2013-09-01

[100 rows x 2 columns]

In [8]:
city_date_group_result = data.groupby("City")["dt"].agg([min, max])
min_city = city_date_group_result["min"].max()
print(min_city)
max_city = city_date_group_result["max"].min()
print(max_city)

1881-03-01
2013-09-01


We see that the latest the data begins from is 1881 and all data points end in Sep'2013.

1st Task: a graphical visualization of the change in temperatures over time, highlighting the cities where the largest temperature ranges were recorded during different historical periods.

Converting date format:

In [9]:
processor = Processor()
processed_data = processor.transform_data(data)

In [10]:
cities_anomaly = processor.calculate_city_temp_ranges(processed_data)

In [11]:
top_cities_anomaly = processor.find_top_anomaly_cities(cities_anomaly)
top_cities_anomaly

City  Year     min     max  temp_range
34     Harbin  1922 -26.772  23.148      49.920
19  Changchun  1919 -20.179  26.572      46.751
64   Montreal  1761 -15.568  26.164      41.732
85   Shenyang  1922 -17.035  23.925      40.960
65     Moscow  2010 -15.846  24.580      40.426